feat(errors): Rebuild errors table

lynnagara opened the pull request:
Adds a migration that rebuilds the errors table, with a number of
schema changes including:
- Removed columns: org_id, event_string, event_hash, primary_hex_hash,
_tags_flattened and _contexts_flattened
- primary_hash changed from FixedString32 to UUID
- Primary key changed to `(project_id, toStartOfDay(timestamp), primary_hash, cityHash64(event_id))`
- Sharding key for distributed table has changed
- Partitioning key has changed to allow values other than 30 or 90

Since we aren't reading anything from this table yet, we are
just going to drop the old table and any data that was in it.

Before merging this, we should also:
- stop any running errors consumer (and replacer) due to changes in the
errors processor
- https://github.com/getsentry/snuba/pull/1353 as running the old
system's migration after this one could cause problems

fpacifici commented on snuba/datasets/errors_processor.py:
```diff
@@ -100,7 +99,7 @@ def extract_contexts_custom(
         metadata: KafkaMessageMetadata,
     ) -> None:
-        output["_contexts_flattened"] = ""
+        pass
```
I'd just remove the method from the parent class. None is supposed to use this now.

fpacifici commented on snuba/migrations/snuba_migrations/events/0011_rebuild_errors.py:
```diff
@@ -0,0 +1,185 @@
+from snuba.migrations import migration, operations, table_engines
+
+
+columns = [
```
I'd assume you agreed on this schema with James. Is that correct ?

fpacifici commented on snuba/migrations/snuba_migrations/events/0011_rebuild_errors.py:
```diff
@@ -0,0 +1,185 @@
+                    local_table_name="errors_local", sharding_key=sample_expr,
+                ),
+            ),
+            operations.DropTable(
```
There won't be a _tags_hash_map column on the distributed table this way. Is that intentional ?

fpacifici reviewed (changes requested):
I think the distributed table is missing the _tags_hash_map column.

fpacifici commented on snuba/migrations/snuba_migrations/events/0011_rebuild_errors.py:
```diff
@@ -0,0 +1,200 @@
+        - no org_id
+        - cityHash64(toString(event_id)) instead of primary_hash_hex, event_hash
+    - Sharding key for distributed table is also cityHash64(toString(event_id))
+    - No _tags_flattened and _contexts_flattened since we are mostly no longer
+    using these already
```
They are not used at all.

lynnagara commented on snuba/migrations/snuba_migrations/events/0011_rebuild_errors.py:
```diff
@@ -0,0 +1,200 @@
+        ]
+
+    def forwards_dist(self) -> Sequence[operations.Operation]:
+        def strip_materialized(columns: Sequence[Column]) -> None:
+            for col in columns:
+                if isinstance(col.type, Materialized):
+                    col.type = col.type.inner_type
+
+        dist_columns = deepcopy(columns)
+        strip_materialized(dist_columns)
```
Interesting, I guess this has been fixed in ClickHouse for quite a while now so I will remove it. (I am actually not sure which ClickHouse versions we officially support)

fpacifici commented on snuba/datasets/storages/errors.py:
```diff
@@ -69,9 +64,7 @@
         ("retention_days", UInt(16)),
         ("deleted", UInt(8)),
```
We are likely going to need to turn this into a real version column. 
I think we should rename it to row_version (version already exists), or soemething else while we are rebuilding the table.

lynnagara replied:
I'll go ahead and rename it - two questions
1 - Is UInt8 sufficient or do you want to increase the size?
2 - What value do you want the replacer to write for a deletion?

lynnagara replied:
@fpacifici I went ahead with writing 255 instead of 1 for deletions - does that work for you?

lynnagara commented:
As discussed offline, we are going to keep the deleted column as-is, rather than introducing a version (or row_version) at this point. This is similar to how replacements work today - we rely on rows that are added later with the same `deleted` value to replace earlier rows.

lynnagara commented:
This is the DDL generated from the migration system:

```
CREATE TABLE IF NOT EXISTS errors_local_new (project_id UInt64, timestamp DateTime, event_id UUID CODEC (NONE), platform LowCardinality(String), environment LowCardinality(Nullable(String)), release LowCardinality(Nullable(String)), dist LowCardinality(Nullable(String)), ip_address_v4 Nullable(IPv4), ip_address_v6 Nullable(IPv6), user String DEFAULT '', user_hash UInt64 MATERIALIZED cityHash64(user), user_id Nullable(String), user_name Nullable(String), user_email Nullable(String), sdk_name LowCardinality(Nullable(String)), sdk_version LowCardinality(Nullable(String)), http_method LowCardinality(Nullable(String)), http_referer Nullable(String), tags Nested(key String, value String), contexts Nested(key String, value String), transaction_name LowCardinality(String) DEFAULT '', transaction_hash UInt64 MATERIALIZED cityHash64(transaction_name), span_id Nullable(UInt64), trace_id Nullable(UUID), partition UInt16, offset UInt64 CODEC (DoubleDelta, LZ4), message_timestamp DateTime, retention_days UInt16, deleted UInt8, group_id UInt64, primary_hash UUID, received DateTime, message String, title String, culprit String, level LowCardinality(String), location Nullable(String), version LowCardinality(Nullable(String)), type LowCardinality(String), exception_stacks Nested(type Nullable(String), value Nullable(String), mechanism_type Nullable(String), mechanism_handled Nullable(UInt8)), exception_frames Nested(abs_path Nullable(String), colno Nullable(UInt32), filename Nullable(String), function Nullable(String), lineno Nullable(UInt32), in_app Nullable(UInt8), package Nullable(String), module Nullable(String), stack_level Nullable(UInt16)), sdk_integrations Array(String), modules Nested(name String, version String)) ENGINE ReplacingMergeTree(deleted) ORDER BY (project_id, toStartOfDay(timestamp), primary_hash, cityHash64(event_id)) PARTITION BY (retention_days, toMonday(timestamp)) SAMPLE BY cityHash64(event_id) TTL timestamp + toIntervalDay(retention_days) SETTINGS index_granularity=8192;
ALTER TABLE errors_local_new ADD COLUMN IF NOT EXISTS _tags_hash_map Array(UInt64) MATERIALIZED arrayMap((k, v) -> cityHash64(concat(replaceRegexpAll(k, '(\\=|\\\\)', '\\\\\\1'), '=', v)), tags.key, tags.value) AFTER tags;
DROP TABLE IF EXISTS errors_local;
RENAME TABLE errors_local_new TO errors_local;
CREATE TABLE IF NOT EXISTS errors_dist_new (project_id UInt64, timestamp DateTime, event_id UUID CODEC (NONE), platform LowCardinality(String), environment LowCardinality(Nullable(String)), release LowCardinality(Nullable(String)), dist LowCardinality(Nullable(String)), ip_address_v4 Nullable(IPv4), ip_address_v6 Nullable(IPv6), user String DEFAULT '', user_hash UInt64 MATERIALIZED cityHash64(user), user_id Nullable(String), user_name Nullable(String), user_email Nullable(String), sdk_name LowCardinality(Nullable(String)), sdk_version LowCardinality(Nullable(String)), http_method LowCardinality(Nullable(String)), http_referer Nullable(String), tags Nested(key String, value String), contexts Nested(key String, value String), transaction_name LowCardinality(String) DEFAULT '', transaction_hash UInt64 MATERIALIZED cityHash64(transaction_name), span_id Nullable(UInt64), trace_id Nullable(UUID), partition UInt16, offset UInt64 CODEC (DoubleDelta, LZ4), message_timestamp DateTime, retention_days UInt16, deleted UInt8, group_id UInt64, primary_hash UUID, received DateTime, message String, title String, culprit String, level LowCardinality(String), location Nullable(String), version LowCardinality(Nullable(String)), type LowCardinality(String), exception_stacks Nested(type Nullable(String), value Nullable(String), mechanism_type Nullable(String), mechanism_handled Nullable(UInt8)), exception_frames Nested(abs_path Nullable(String), colno Nullable(UInt32), filename Nullable(String), function Nullable(String), lineno Nullable(UInt32), in_app Nullable(UInt8), package Nullable(String), module Nullable(String), stack_level Nullable(UInt16)), sdk_integrations Array(String), modules Nested(name String, version String)) ENGINE Distributed(cluster_name, default, errors_local, cityHash64(event_id));
ALTER TABLE errors_dist_new ADD COLUMN IF NOT EXISTS _tags_hash_map Array(UInt64) AFTER tags;
DROP TABLE IF EXISTS errors_dist;
RENAME TABLE errors_dist_new TO errors_dist;
CREATE TABLE IF NOT EXISTS errors_dist_ro (project_id UInt64, timestamp DateTime, event_id UUID CODEC (NONE), platform LowCardinality(String), environment LowCardinality(Nullable(String)), release LowCardinality(Nullable(String)), dist LowCardinality(Nullable(String)), ip_address_v4 Nullable(IPv4), ip_address_v6 Nullable(IPv6), user String DEFAULT '', user_hash UInt64 MATERIALIZED cityHash64(user), user_id Nullable(String), user_name Nullable(String), user_email Nullable(String), sdk_name LowCardinality(Nullable(String)), sdk_version LowCardinality(Nullable(String)), http_method LowCardinality(Nullable(String)), http_referer Nullable(String), tags Nested(key String, value String), contexts Nested(key String, value String), transaction_name LowCardinality(String) DEFAULT '', transaction_hash UInt64 MATERIALIZED cityHash64(transaction_name), span_id Nullable(UInt64), trace_id Nullable(UUID), partition UInt16, offset UInt64 CODEC (DoubleDelta, LZ4), message_timestamp DateTime, retention_days UInt16, deleted UInt8, group_id UInt64, primary_hash UUID, received DateTime, message String, title String, culprit String, level LowCardinality(String), location Nullable(String), version LowCardinality(Nullable(String)), type LowCardinality(String), exception_stacks Nested(type Nullable(String), value Nullable(String), mechanism_type Nullable(String), mechanism_handled Nullable(UInt8)), exception_frames Nested(abs_path Nullable(String), colno Nullable(UInt32), filename Nullable(String), function Nullable(String), lineno Nullable(UInt32), in_app Nullable(UInt8), package Nullable(String), module Nullable(String), stack_level Nullable(UInt16)), sdk_integrations Array(String), modules Nested(name String, version String)) ENGINE Distributed(cluster_name, default, errors_local, cityHash64(event_id));
```
